File: functions/src/models/ThemeConfig.ts
export interface ThemeConfig {
  colors: {
    primary: string;
    secondary: string;
    background: string;
    surface: string;
    error: string;
    success: string;
    warning: string;
    info: string;
    invoice: string;
    action: string;
    link: string;
    textPrimary: string;
    textSecondary: string;
    textHint: string;
    cardBackground: string;
    cardText: string;
  };

  typography: {
    fontFamily: string;
    fontSizes: {
      h1: number;
      h2: number;
      h3: number;
      body: number;
      caption: number;
      button: number;
    };
    fontWeights: {
      light: string;
      regular: string;
      medium: string;
      semibold: string;
      bold: string;
    };
    lineHeights: {
      tight: number;
      normal: number;
      relaxed: number;
    };
  };

  spacing: {
    xs: number;
    sm: number;
    md: number;
    lg: number;
    xl: number;
    xxl: number;
  };

  borderRadius: {
    none: number;
    sm: number;
    md: number;
    lg: number;
    xl: number;
    full: number;
  };

  shadows: {
    enabled: boolean;
    sm: string;
    md: string;
    lg: string;
    xl: string;
  };

  effects: {
    enableGlassmorphism: boolean;
    enableGradients: boolean;
    enableAnimations: boolean;
    animationSpeed: "slow" | "normal" | "fast";
    glassOpacity: number;
    glassBlur: number;
  };

  components: {
    buttons: {
      borderRadius: number;
      elevation: number;
      padding: { horizontal: number; vertical: number };
    };
    cards: {
      borderRadius: number;
      elevation: number;
      padding: number;
      enableBorder: boolean;
      borderColor?: string;
      borderWidth?: number;
    };
    inputs: {
      borderRadius: number;
      height: number;
      padding: number;
      borderWidth: number;
      borderColor: string;
      focusBorderColor: string;
    };
    appBar: {
      height: number;
      elevation: number;
      backgroundColor?: string;
      centerTitle: boolean;
    };
    bottomNav: {
      height: number;
      elevation: number;
      backgroundColor?: string;
      selectedItemColor?: string;
      unselectedItemColor?: string;
      showLabels: boolean;
    };
  };

  darkMode: {
    enabled: boolean;
    defaultMode: "light" | "dark" | "auto";
    followSystem: boolean;
    darkColors?: {
      primary?: string;
      secondary?: string;
      background?: string;
      surface?: string;
      textPrimary?: string;
      textSecondary?: string;
    };
  };
}

export const defaultThemeConfig: ThemeConfig = {
  colors: {
    primary: "#673AB7",
    secondary: "#9575CD",
    background: "#0F172A",
    surface: "#1E293B",
    error: "#EF4444",
    success: "#10B981",
    warning: "#F59E0B",
    info: "#3B82F6",
    invoice: "#10B981",
    action: "#E11D48",
    link: "#3B82F6",
    textPrimary: "#FFFFFF",
    textSecondary: "#94A3B8",
    textHint: "#64748B",
    cardBackground: "#F8F8F8",
    cardText: "#333333",
  },
  typography: {
    fontFamily: "Inter",
    fontSizes: { h1: 32, h2: 24, h3: 20, body: 14, caption: 12, button: 14 },
    fontWeights: { light: "300", regular: "400", medium: "500", semibold: "600", bold: "700" },
    lineHeights: { tight: 1.2, normal: 1.5, relaxed: 1.75 },
  },
  spacing: { xs: 4, sm: 8, md: 16, lg: 24, xl: 32, xxl: 48 },
  borderRadius: { none: 0, sm: 4, md: 8, lg: 16, xl: 24, full: 9999 },
  shadows: {
    enabled: true,
    sm: "0 1px 2px 0 rgba(0, 0, 0, 0.05)",
    md: "0 4px 6px -1px rgba(0, 0, 0, 0.1)",
    lg: "0 10px 15px -3px rgba(0, 0, 0, 0.1)",
    xl: "0 20px 25px -5px rgba(0, 0, 0, 0.1)",
  },
  effects: {
    enableGlassmorphism: true,
    enableGradients: true,
    enableAnimations: true,
    animationSpeed: "normal",
    glassOpacity: 0.7,
    glassBlur: 12,
  },
  components: {
    buttons: { borderRadius: 18, elevation: 2, padding: { horizontal: 24, vertical: 14 } },
    cards: { borderRadius: 16, elevation: 2, padding: 16, enableBorder: false, borderColor: "#E2E8F0", borderWidth: 1 },
    inputs: { borderRadius: 12, height: 48, padding: 12, borderWidth: 1, borderColor: "#E2E8F0", focusBorderColor: "#673AB7" },
    appBar: { height: 56, elevation: 0, backgroundColor: undefined, centerTitle: true },
    bottomNav: { height: 74, elevation: 8, backgroundColor: undefined, selectedItemColor: undefined, unselectedItemColor: undefined, showLabels: false },
  },
  darkMode: {
    enabled: true,
    defaultMode: "auto",
    followSystem: true,
    darkColors: { background: "#0F172A", surface: "#1E293B", textPrimary: "#F8FAFC", textSecondary: "#94A3B8" },
  },
};

export function isValidHexColor(color: string): boolean {
  return /^#[0-9A-F]{6}$/i.test(color);
}

export function validateThemeConfig(config: Partial<ThemeConfig>): { valid: boolean; errors: string[] } {
  const errors: string[] = [];
  
  if (config.colors) {
    Object.entries(config.colors).forEach(([key, value]) => {
      if (value && !isValidHexColor(value)) {
        errors.push(`Cor inválida em colors.${key}: ${value}`);
      }
    });
  }
  
  if (config.typography?.fontSizes) {
    Object.entries(config.typography.fontSizes).forEach(([key, value]) => {
      if (value && (value < 8 || value > 72)) {
        errors.push(`Tamanho de fonte inválido em typography.fontSizes.${key}: ${value}`);
      }
    });
  }
  
  if (config.spacing) {
    Object.entries(config.spacing).forEach(([key, value]) => {
      if (value && (value < 0 || value > 200)) {
        errors.push(`Espaçamento inválido em spacing.${key}: ${value}`);
      }
    });
  }
  
  if (config.effects?.glassOpacity !== undefined) {
    if (config.effects.glassOpacity < 0 || config.effects.glassOpacity > 1) {
      errors.push(`Opacidade inválida: ${config.effects.glassOpacity}`);
    }
  }
  
  return { valid: errors.length === 0, errors };
}

export function mergeThemeConfig(custom: Partial<ThemeConfig>, defaults: ThemeConfig = defaultThemeConfig): ThemeConfig {
  return {
    colors: { ...defaults.colors, ...(custom.colors || {}) },
    typography: {
      ...defaults.typography,
      ...(custom.typography || {}),
      fontSizes: { ...defaults.typography.fontSizes, ...(custom.typography?.fontSizes || {}) },
      fontWeights: { ...defaults.typography.fontWeights, ...(custom.typography?.fontWeights || {}) },
      lineHeights: { ...defaults.typography.lineHeights, ...(custom.typography?.lineHeights || {}) },
    },
    spacing: { ...defaults.spacing, ...(custom.spacing || {}) },
    borderRadius: { ...defaults.borderRadius, ...(custom.borderRadius || {}) },
    shadows: { ...defaults.shadows, ...(custom.shadows || {}) },
    effects: { ...defaults.effects, ...(custom.effects || {}) },
    components: {
      buttons: { ...defaults.components.buttons, ...(custom.components?.buttons || {}) },
      cards: { ...defaults.components.cards, ...(custom.components?.cards || {}) },
      inputs: { ...defaults.components.inputs, ...(custom.components?.inputs || {}) },
      appBar: { ...defaults.components.appBar, ...(custom.components?.appBar || {}) },
      bottomNav: { ...defaults.components.bottomNav, ...(custom.components?.bottomNav || {}) },
    },
    darkMode: {
      ...defaults.darkMode,
      ...(custom.darkMode || {}),
      darkColors: { ...(defaults.darkMode.darkColors || {}), ...(custom.darkMode?.darkColors || {}) },
    },
  };
}

File: codigos antigos/Layout 02/lib/models/theme_config.dart
import 'package:flutter/material.dart';

class ThemeConfig {
  final ThemeColors colors;
  final ThemeTypography typography;
  final ThemeSpacing spacing;
  final ThemeBorderRadius borderRadius;
  final ThemeEffects effects;
  final DarkModeConfig darkMode;

  const ThemeConfig({
    required this.colors,
    required this.typography,
    required this.spacing,
    required this.borderRadius,
    required this.effects,
    required this.darkMode,
  });

  factory ThemeConfig.fromJson(Map<String, dynamic> json) {
    return ThemeConfig(
      colors: ThemeColors.fromJson(json['colors'] ?? {}),
      typography: ThemeTypography.fromJson(json['typography'] ?? {}),
      spacing: ThemeSpacing.fromJson(json['spacing'] ?? {}),
      borderRadius: ThemeBorderRadius.fromJson(json['borderRadius'] ?? {}),
      effects: ThemeEffects.fromJson(json['effects'] ?? {}),
      darkMode: DarkModeConfig.fromJson(json['darkMode'] ?? {}),
    );
  }

  static ThemeConfig get defaultTheme => ThemeConfig(
        colors: ThemeColors.defaultColors,
        typography: ThemeTypography.defaultTypography,
        spacing: ThemeSpacing.defaultSpacing,
        borderRadius: ThemeBorderRadius.defaultBorderRadius,
        effects: ThemeEffects.defaultEffects,
        darkMode: DarkModeConfig.defaultDarkMode,
      );
}

class ThemeColors {
  final Color primary;
  final Color secondary;
  final Color background;
  final Color surface;
  final Color error;
  final Color success;
  final Color warning;
  final Color info;
  final Color textPrimary;
  final Color textSecondary;
  final Color textHint;

  const ThemeColors({
    required this.primary,
    required this.secondary,
    required this.background,
    required this.surface,
    required this.error,
    required this.success,
    required this.warning,
    required this.info,
    required this.textPrimary,
    required this.textSecondary,
    required this.textHint,
  });

  factory ThemeColors.fromJson(Map<String, dynamic> json) {
    return ThemeColors(
      primary: _parseColor(json['primary'], const Color(0xFF673AB7)),
      secondary: _parseColor(json['secondary'], const Color(0xFF9575CD)),
      background: _parseColor(json['background'], const Color(0xFF0F172A)),
      surface: _parseColor(json['surface'], const Color(0xFF1E293B)),
      error: _parseColor(json['error'], const Color(0xFFEF4444)),
      success: _parseColor(json['success'], const Color(0xFF10B981)),
      warning: _parseColor(json['warning'], const Color(0xFFF59E0B)),
      info: _parseColor(json['info'], const Color(0xFF3B82F6)),
      textPrimary: _parseColor(json['textPrimary'], Colors.white),
      textSecondary: _parseColor(json['textSecondary'], const Color(0xFF94A3B8)),
      textHint: _parseColor(json['textHint'], const Color(0xFF64748B)),
    );
  }

  static Color _parseColor(dynamic value, Color fallback) {
    if (value == null) return fallback;
    if (value is String && value.startsWith('#')) {
      try {
        return Color(int.parse(value.substring(1), radix: 16) + 0xFF000000);
      } catch (_) {
        return fallback;
      }
    }
    return fallback;
  }

  static const ThemeColors defaultColors = ThemeColors(
    primary: Color(0xFF673AB7),
    secondary: Color(0xFF9575CD),
    background: Color(0xFF0F172A),
    surface: Color(0xFF1E293B),
    error: Color(0xFFEF4444),
    success: Color(0xFF10B981),
    warning: Color(0xFFF59E0B),
    info: Color(0xFF3B82F6),
    textPrimary: Colors.white,
    textSecondary: Color(0xFF94A3B8),
    textHint: Color(0xFF64748B),
  );
}

class ThemeTypography {
  final String fontFamily;
  final Map<String, double> fontSizes;

  const ThemeTypography({required this.fontFamily, required this.fontSizes});

  factory ThemeTypography.fromJson(Map<String, dynamic> json) {
    final sizes = json['fontSizes'] as Map<String, dynamic>? ?? {};
    return ThemeTypography(
      fontFamily: json['fontFamily'] ?? 'Inter',
      fontSizes: {
        'h1': (sizes['h1'] ?? 32).toDouble(),
        'h2': (sizes['h2'] ?? 24).toDouble(),
        'h3': (sizes['h3'] ?? 20).toDouble(),
        'body': (sizes['body'] ?? 14).toDouble(),
        'caption': (sizes['caption'] ?? 12).toDouble(),
      },
    );
  }

  static const ThemeTypography defaultTypography = ThemeTypography(
    fontFamily: 'Inter',
    fontSizes: {'h1': 32, 'h2': 24, 'h3': 20, 'body': 14, 'caption': 12},
  );
}

class ThemeSpacing {
  final double xs;
  final double sm;
  final double md;
  final double lg;
  final double xl;

  const ThemeSpacing({required this.xs, required this.sm, required this.md, required this.lg, required this.xl});

  factory ThemeSpacing.fromJson(Map<String, dynamic> json) {
    return ThemeSpacing(
      xs: (json['xs'] ?? 4).toDouble(),
      sm: (json['sm'] ?? 8).toDouble(),
      md: (json['md'] ?? 16).toDouble(),
      lg: (json['lg'] ?? 24).toDouble(),
      xl: (json['xl'] ?? 32).toDouble(),
    );
  }

  static const ThemeSpacing defaultSpacing = ThemeSpacing(xs: 4, sm: 8, md: 16, lg: 24, xl: 32);
}

class ThemeBorderRadius {
  final double sm;
  final double md;
  final double lg;
  final double xl;

  const ThemeBorderRadius({required this.sm, required this.md, required this.lg, required this.xl});

  factory ThemeBorderRadius.fromJson(Map<String, dynamic> json) {
    return ThemeBorderRadius(
      sm: (json['sm'] ?? 4).toDouble(),
      md: (json['md'] ?? 8).toDouble(),
      lg: (json['lg'] ?? 16).toDouble(),
      xl: (json['xl'] ?? 24).toDouble(),
    );
  }

  static const ThemeBorderRadius defaultBorderRadius = ThemeBorderRadius(sm: 4, md: 8, lg: 16, xl: 24);
}

class ThemeEffects {
  final bool enableGlassmorphism;
  final bool enableGradients;
  final bool enableAnimations;
  final double glassOpacity;
  final double glassBlur;

  const ThemeEffects({
    required this.enableGlassmorphism,
    required this.enableGradients,
    required this.enableAnimations,
    required this.glassOpacity,
    required this.glassBlur,
  });

  factory ThemeEffects.fromJson(Map<String, dynamic> json) {
    return ThemeEffects(
      enableGlassmorphism: json['enableGlassmorphism'] ?? true,
      enableGradients: json['enableGradients'] ?? true,
      enableAnimations: json['enableAnimations'] ?? true,
      glassOpacity: (json['glassOpacity'] ?? 0.7).toDouble(),
      glassBlur: (json['glassBlur'] ?? 12).toDouble(),
    );
  }

  static const ThemeEffects defaultEffects = ThemeEffects(
    enableGlassmorphism: true,
    enableGradients: true,
    enableAnimations: true,
    glassOpacity: 0.7,
    glassBlur: 12,
  );
}

class DarkModeConfig {
  final bool enabled;
  final String defaultMode;

  const DarkModeConfig({required this.enabled, required this.defaultMode});

  factory DarkModeConfig.fromJson(Map<String, dynamic> json) {
    return DarkModeConfig(
      enabled: json['enabled'] ?? true,
      defaultMode: json['defaultMode'] ?? 'auto',
    );
  }

  static const DarkModeConfig defaultDarkMode = DarkModeConfig(enabled: true, defaultMode: 'auto');
}
